rfc/src/rfcs/0010-rust-ub-checks.md

```diff
@@ -0,0 +1,175 @@
+
+## Software Design
+
+*To be filled in the next iteration*
+
+## Rationale and alternatives
+
+Kani already has undefined behavior checks, some are implemented on the CBMC side, such as out-of-bounds access, and
+arithmetic overflow.
+While others are implemented on Kani's side, such as intrinsics validation checks, and some are a
+result of instrumentation to the Rust standard library.
+
+The behavior of these checks varies according to how they are added.
+The ones on the CBMC side do not interrupt the analysis of a path, while the ones on the Kani / Rust std side will.
+
+For example, memory bound checks are implemented in CBMC, and the following harness today will trigger 2 different
+failures:
+
+```rust
+#[kani::proof]
+pub fn is_zero() {
+    let var = [0u32; 4];
+    let ptr = &var as *const u32;
+    let idx: usize = 4;
+    assert_eq!(unsafe { ptr.add(idx).read() }, 0);
+}
+```
+
+The failed checks are:
+
+```plaintext
+Failed Checks: assertion failed: unsafe { ptr.add(idx).read() } == 0
+Failed Checks: dereference failure: pointer outside object bounds
+```
+
+While the following harness only fails the standard library check:
+
+```rust
+#[kani::proof]
+pub fn is_zero() {
+    let var = [0; 4];
+    let idx: usize = 4;
+    assert_eq!(unsafe { *var.get_unchecked(idx) }, 0);
+}
+```
+
+Failing due to an assumption check inside `get_unchecked` function.
+
+```plaintext
+Failed Checks: assumption failed
+```
+
+We would like to keep taking advantage of existing instrumentation to maximize the UB check coverage.
+The mechanism used to build the check, however, shouldn't influence the user experience.
+
+### Opt-out vs opt-in
+
+The default behavior matches the one that we believe is the recommended for most users due to soundness.
+In some cases, users may need to disable some checks for reasons such as spurious CEX or proof scalability.
+Thus, we suggest that every category of checks should have an opt-out mechanism that can either be global or local;
+i.e.: via command argument or harness attribute.
+
+The main disadvantage of this model is that stabilizing a new check category is a non-backward compatible change.
+I.e.: It may impact the status of a previously passing harness if a new check fails.
+
+## Open questions
+
+- Should we add a new status for potentially spurious counter examples? Such as:
+    - Counter examples detected together with a non-deterministic UB check.
+    - Over-approximating UB checks.
+- How should UB checks affect concrete playback? Is a comment enough?
+    - Another possibility would be to add a `unreachable!()` statement at the end of the test case with a message
+      explaining why the statement should be unreachable.
+- Should we deprecate the `--no-default-checks` and all `--[CHECK-NAME]-checks` checks.
+
+## Out of scope / Future Improvements
+
+- **List enabled checks:** Kani could include a command that list all the categories of undefined behaviors supported.
```

Comment:
I'd like to have a table (generated from the source code?) which is available in our online documentation.

Reply:
I would love that too. :smiley: